src/ansi-c/ansi_c_convert_type.cpp

```diff
@@ -253,26 +254,22 @@ void ansi_c_convert_typet::read_rec(const typet &type)
     const exprt &as_expr =
       static_cast<const exprt &>(static_cast<const irept &>(type));
 
-    for(const exprt &operand : to_unary_expr(as_expr).op().operands())
+    for(const exprt &target : to_unary_expr(as_expr).op().operands())
     {
-      if(
-        operand.id() != ID_symbol && operand.id() != ID_ptrmember &&
-        operand.id() != ID_member && operand.id() != ID_dereference)
+      if(!is_lvalue(target))
       {
-        error().source_location = source_location;
+        error().source_location = target.source_location();
         error() << "illegal target in assigns clause" << eom;
         throw 0;
       }
+      else if(has_subexpr(target, ID_side_effect))
+      {
+        error().source_location = target.source_location();
+        error() << "Assigns clause is not side-effect free." << eom;
```

Comment:
Definitely add a test.

Reply:
This is actually not necessary, because we check for `is_lvalue`prior, which would filter anything with side effects. Take a look at this example.
```c
#include <assert.h>
#include <stdlib.h>

int x = 0;

void bar(int *y)
{
  y = x;
}

void foo(int *y, int z) __CPROVER_assigns((x == 0) ? bar(y) : x)
{
  *y = z;
  z = -2;
}

int main()
{
  int *y = malloc(sizeof(*y));
  *y = 0;
  int z = 1;
  foo(y, z);
  assert(z == 1);
  return 0;
}
```
Verification  yields to this result.
```
main.c:11:1: error: illegal target in assigns clause
 void foo(int *y, int z) __CPROVER_assigns((x == 0) ? bar(y) : x)
CONVERSION ERROR
```

Are there any cases we might miss if we just remove this? @SaswatPadhi @jimgrundy

Reply:
have we tried "__CPROVER_assigns(++x)" or similar?

Reply:
Also, coverage report says this isn't covered. If it is hittable then we should have a test, just need to think of one. If it isn't hittable then this code is dead and should be removed.

Reply:
> This is actually not necessary, because we check for `is_lvalue` prior, which would filter anything with side effects. 

I don't think so. In the example you tried, `bar(y)` would return `void` and so may be that's why it's not treated as a valid l-value. But, in general, function calls are valid within l-values.

Additionally, I don't think expressions like `*(++x)` would be flagged by `is_lvalue`. We should add a regression test with such a target in the assigns clause and make sure that we hit this branch in code.